doc/adr/0010-marconi-indexer.rst

```diff
@@ -0,0 +1,133 @@
+
+At this point all query results are synchronous. We have plans to extend this functionality, but these plans are based on updating the notification system for indexers which will be described in a further ADR.
+
+API Design
+^^^^^^^^^^
+
+We want the API for our users to be as flexible as possible, so some of our previously mentioned design patterns are not captured by the API, but rather by its implementation.
+
+Data types
+""""""""""
+
+* The `Events` data type contains the following fields:
+
+  * Slot numbers (a data type that supports ordering)
+  * Block id (a data type that supports equality checking)
+  * e (type variable standing for the event)
+
+* The `Query` data type contains the following fields:
+
+  * Validity interval (can be any interval defined by using the slot numbers or a special value that turn off checking for validity)
+  * q (type variable standing for the query)
+
+* Result
+
+  * Slot number at which the query was ran
+  * r (type variable standing for the query result)
+
+Functions
+"""""""""
+
+* The `Query` function takes the following parameters:
+
+  * Indexer - The indexer that we are using to run the query.
+  * Validity interval - The interval under which the query needs to be ran.
+  * The query (q type variable) - The user-defined query.
+  * The query result
+
+* The `Store` function takes the following parameters:
+
+  * Indexer - The indexer for which we run the function
+  * Does not return anything useful
+
+* The `Resume` function takes the following parameters:
+
+  * Indexer - The indexer we need to query for the last consumed slot numbers
+  * Returns a list of slot numbers and block ids
+
+Runtime parameters
+""""""""""""""""""
+
+* Minimum events retained (this should be the previously mentioned K parameter)
+
+  Currently the main cardano network guarantees that there will be no rollbacks beyond 2160 blocks. This would be that parameter.
+
+* Maximum in-memory events (should be less than K)
+
+  How many events to we want to keep in memory. For the main network this should be any number lower that 2160. The larger the number the less frequent writing to disk is and the more RAM is used.
+
+Extension mechanisms
+""""""""""""""""""""
+
+A. Storage engine
+
+You can customise the query and store functions which run in some generic monad to use whatever backend is best for the job. We currently use SQLite, but that is more for convenience than anything else.
+
+B. Query intervals
+
+If you want to specify an interval for your queries (which is highly encouraged) then you need to have either in memory or on disk sufficient information to reconstruct the state at the given slot number. By storing more than K events you can extend the query interval as much as you need. In extreme, you can only store events on disk, in which case your queries can span the whole blockchain.
```

Comment:
I don't quite understand what your intended meaning is for this sentence: 

> In extreme, you can only store events on disk, in which case your queries can span the whole blockchain.

If I am understanding your meaning correctly, I would suggest this wording: 

> In the extreme case in which you can only store events on disk, your queries can span the whole blockchain.

Reply:
I understand what you are saying more clearly now.